### PR TITLE
Writing a paper for JOSS.

### PR DESCRIPTION
**Pull request by daffidwilde**
Write a paper to accompany this library that is to be submitted to the [Journal of Open Source Education (JOSE)](https://jose.theoj.org). Instructions and requirements for submission are available in the [documentation](https://openjournals.readthedocs.io/en/jose/). In summary, the package and paper must:

1. Be "open-source software, created as educational technology or infrastructure". In particular:
   > Software submissions should make a clear contribution to the available open-source software that supports teaching and learning, or makes an educational process better (e.g., faster, easier, simpler).

2. Be fully [open](http://opendefinition.org/) and available (hosted on GitHub with CC-BY and OSI-approved licenses for text and code respectively)

3. Contain a short (approx. 1000 words) `paper.md` file that includes the following:
   - List all authors and affiliations.
   - Describe the submission, and explain its eligibility for JOSE.
   - Include a “Statement of Need” section, explaining how the submitted artefacts contribute to computationally enabled teaching and learning, and describing how they might be adopted by others.
   - Describe the functionality of the software, usage and recent experience of use in teaching and learning situations.
   - Tell the “story” of the project: how did it come to be?
   - Cite key references, including a link to the open archive of the software or the learning module.

### CODE REVIEW

**Comment by daffidwilde** [comment]
Following thought and conversation with the co-authors, this paper will be aimed at JOSS rather than JOSE.

The principles of the paper remain largely the same except the focus moves from education to the software itself.

File: paper.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+# Summary
+
+Matching games allow for the allocation of resources and partnerships in a fair
+way. Typically, a matching game is defined by two sets of players that each have
+preferences over at least some of the elements of the other set. The objective
+of the game is then to find a mapping between the sets of players in which
+everyone is *happy enough* with their match.
+
+One of the most ubiquitous matching games models the Stable Marriage Problem
```
Comment (drvinceknight) [inline paper.md]: `One of the most ubiquitous matching game models is the Stable Marriage Problem (SM)`

File: paper.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+(SM). In SM, there are two distinct player sets of size $N$: a set of suitors
+$S$ and a set of reviewers $R$. Each suitor must strictly rank all of the
+reviewers, and vice versa. This arrangement of suitors, reviewers, and
+their preferences is called a game of size $N$ [@gale1962].
+
+In SM, a matching is any bijection $M$ between $S$ and $R$, and it is considered
+to be stable (i.e. everyone is happy enough) if it contains no blocking pairs.
```
Comment (drvinceknight) [inline paper.md]: `(i.e. no one has a reason to modify their current match)`

File: paper.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+academia and industry.
+
+Matching is a Python library that relies on one core library from the
+standard scientific Python stack -- NumPy [@numpy] -- that currently implements
+algorithms for four types of matching games:
+
+- The stable marriage problem (SM) [@gale1962];
+- the hospital-resident assignment problem (HR) [@gale1962; @roth1984];
+- the student-project allocation problem (SA) [@abra2007];
+- the stable roommates problem (SR) [@irvi1985].
```
Comment (drvinceknight) [inline paper.md]: Does this differ from `MatchingR`? What does `MatchingR` solve?

File: paper.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+In addition to this, Matching has been developed to a high degree of best
+practices in research software development [@jime2017], and is thoroughly
```
Comment (drvinceknight) [inline paper.md]: `practice`

File: paper.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+>>> game = HospitalResident.create_from_dictionaries(
+...     resident_prefs, hospital_prefs, capacities
+>>> )
+>>> _ = game.solve() # 48.6 ms ± 963 µs per loop
+```
+
+Matching has been designed to be used as a research tool and to aid in the
+education of game theory. It is currently being used by a number of
+undergraduate students and postgraduate researchers in universities around the
+world, and has been used to massively streamline the final year project
```
Comment (drvinceknight) [inline paper.md]: Worth including the k means paper?

File: paper.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+allocation process for one of the largest schools at Cardiff University. With
```
Comment (drvinceknight) [inline paper.md]: `(as a SA problem)`.

**Comment by daffidwilde** [comment]
Fixed @drvinceknight comments in 5303212

**Comment by drvinceknight** [comment]
> Fixed @drvinceknight comments in [5303212](https://github.com/daffidwilde/matching/commit/5303212a88791f421bcdf0659106820d76f4e5c3)

Looks good to me :+1:

**Pull request merged**